### DISCUSSION TITLE
How to send payload with  GenerateSignalUrl()

### DISCUSSION THREAD

**Discussion by ritheshjaston** [elsa-workflows/elsa-core#5626] [Q&A]
var approvalUrl = context.GenerateSignalUrl("ApprovelevelsV4");
 
 In the Above GenerateSignalUrl method I want to add the payload some id or something because the correlation Id will be the same for a workflow 
 
 Code will be Like this
 _______________________
    builder
      .SignalReceived("DocRecievdV4")
      .WithName("Recieving the DocRecievd Signal")
      .Then(context =>
      {
          dynamic input = context.Input;
          var level = input.Approoverlevel;
          context.SetVariable("Approoverdetail", input.Approoverlevel);
          context.SetVariable("counter", 0);
          context.SetVariable("approvalcounter", 0);
          context.SetVariable("Emailcount", level.emails.Count);
          context.SetVariable("emails", level.emails);
          context.SetVariable("maxApprovers", level.maxapprovers);
      })
      .WithName("Getting the Approover")
      .While(context => context.GetVariable<int>("counter") < context.GetVariable<int>("Emailcount"),

                  whileLoop => whileLoop
                   .SendEmail(context => context
                   .WithRecipient(context => context.GetVariable<IList<dynamic>>("emails")[context.GetVariable<int>("counter")].MId)
                   .WithSubject("Approval Process")
                   .WithBody(context =>
                   {
                       var approvalUrl = context.GenerateSignalUrl("ApprovelevelsV4");
                       return $"Document received for review.<br><a href=\"{approvalUrl}\">Approve</a><br>counter: {context.GetVariable<int>("counter")}";
                   })).WithName("Sending Emails To the First Level Approovers")
                   .Then(context =>
                   {
                       var counter = context.GetVariable<int>("counter");
                       context.SetVariable("counter", counter + 1);

                   }
      ))
      .While(context => context.GetVariable<int>("approvalcounter") < context.GetVariable<int>("maxApprovers"),

                  whileLoop => whileLoop
                  .SignalReceived("ApprovelevelsV4")
                  .WriteHttpResponse(context => context
                       .WithStatusCode(HttpStatusCode.OK)
                       .WithContent("Approoved")
                       .WithContentType("text/plain"))
                  .WithName("Recieving the ApprovelevelsV4 Signal")
                  .Then(context =>
                  {
                      Console.WriteLine("Signal Recieved");
                      var counter = context.GetVariable<int>("approvalcounter");
                      context.SetVariable("approvalcounter", counter + 1);
                  })

      )
      .Finish();
}

**Comment by glime-ai** [comment]
To add a payload to the `GenerateSignalUrl` method in Elsa Workflows, you can use query string parameters. The `GenerateSignalUrl` method allows you to specify additional parameters that will be included in the generated URL as query string values. These parameters can then be used to pass data such as an ID or other relevant information that you want to correlate with the signal.

Here's an example of how you can modify the `GenerateSignalUrl` call to include an ID as a query string parameter:

```csharp
.WithBody(context =>
{
    var documentId = context.GetVariable<string>("DocumentId"); // Assuming you have a variable with the document ID
    var approvalUrl = context.GenerateSignalUrl("ApprovelevelsV4", new { id = documentId }); // Add the document ID as a query string parameter
    return $"Document received for review.<br><a href=\"{approvalUrl}\">Approve</a><br>counter: {context.GetVariable<int>("counter")}";
})
```

In this example, the `documentId` is retrieved from the workflow context and included as a query string parameter named `id` in the generated signal URL. When the URL is accessed, the `id` parameter will be available in the HTTP request and can be used to correlate the signal with the specific workflow instance or document.

Please note that the actual method signature and usage might differ based on the version of Elsa Workflows you are using. If you are using Elsa 2, the above example should work. However, if you are using Elsa 3, you might need to adjust the code according to the new API changes.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by ritheshjaston** [reply]
Severity	Code	Description	Project	File	Line	Suppression State
Error (active)	CS1501	No overload for method 'GenerateSignalUrl' takes 2 arguments	Elsa.test.Web	C:\test\tester\src\Elsa.test.Web\Workflow Level\ApprooverlevelV4.cs	95
